### PR TITLE
Adds block downloads and more threads v2

### LINKED ISSUE #264: Index failed blocks w/ retries
Part of the more resilient message handing discussed in #240, will also let a thread retry back-propagation starting at the failed block.

### PR DESCRIPTION
**Pull request by sanderpick**
Closes #264 

### Threads
- Block indexes now have three states:
  - `READY`: payload is downloaded and synced (if created locally)
  - `QUEUED`: waiting to sync
  - `PENDING`: waiting to download
- During the thread traversal, the inner payloads (the encrypted part) are downloaded with a queue (`BlockDownloads`), which are handed in concurrent batches like the other queues, while the thread cursor continues to traverse
- Traversal handles parents concurrently 
- The actual download _could_ be moved to OS-specific background fetch APIs on mobile, similar to sync
- Block nodes now contain two extra links:
  - `target`, pulled out of the encrypted block, enabling arbitrary nesting of data and message blocks
  - `data`, the new home for the data link (previously target)
- With the addition of the extra links, the thread graph is traversable by IPFS without touching the encrypted payloads 🎉, which paves the way for an OrbitDB-like public thread type, and `js-threads`
- **Fixes an invites bug which caused direct invites to never be sent**

### Thread Service
- Logical cleanup, which is good prep for moving threads to their own repo (`go-threads`)

### Cafe Service
- Fixes a bug in `handleDeliverMessage` where the envelope payload was not being correctly unmarshalled

### Files
- Can now take a target (nested files), but disabled until 0.5.0 release

### Messages
- Can now take a target (nested messages), but disabled until 0.5.0 release

### Comments
- deprecated (replaces by messages w/ a target), but still active until 0.5.0 release

TODO:
- [x] Fix up the CLI and API after #821 (files and messages)
- [x] Fix up the the mobile API after #821 (files and messages)
- [x] Fix `FeedItem`, needs an extra field for `data` links (used to be target)
- [x] Add a datastore migration
- [x] Download thread tails in the background

### CODE REVIEW

**Comment by sanderpick** [comment]
Ready for some 👀 here

File: core/api_files.go [hunk 1]
```diff
@@ -78,7 +78,8 @@ func (a *api) addThreadFiles(g *gin.Context) {
 		return
 	}
 
-	hash, err := thrd.AddFiles(node, opts["caption"], keys.Files)
+	// @todo Allow the setting of the target in 0.5.0
+	hash, err := thrd.AddFiles(node, "", opts["caption"], keys.Files)
```
Comment (carsonfarmer) [inline core/api_files.go]: Nice! Looking forward to that :)

File: core/cafes_test.go [hunk 2]
```diff
@@ -136,12 +136,12 @@ func TestCore_HandleCafeRequests(t *testing.T) {
 
 	// check if blocks are pinned
 	var blocks []string
-	var targets []string
+	var datas []string
```
Comment (carsonfarmer) [inline core/cafes_test.go]: 🤓

File: core/invites.go [hunk 3]
```diff
@@ -211,40 +211,46 @@ func (t *Textile) handleThreadAdd(plaintext []byte, parents []string) (mh.Multih
...
-		return nil, err
-	}
-	hash, err := thrd.join(author)
+	hash, err := thread.join(block.Header.Author)
 	if err != nil {
 		return nil, err
 	}
+
+	// handle the thread tail in the background
+	go func() {
```
Comment (carsonfarmer) [inline core/invites.go]: nice!

File: core/main_test.go [hunk 4]
```diff
@@ -33,7 +33,7 @@ func TestInitRepo(t *testing.T) {
 	if err := InitRepo(InitConfig{
 		Account:  accnt,
 		RepoPath: vars.repoPath,
-		ApiAddr:  fmt.Sprintf("127.0.0.1:%s", GetRandomPort()),
+		ApiAddr:  fmt.Sprintf("127.0.0.1:9999"),
```
Comment (carsonfarmer) [inline core/main_test.go]: We going to do this throughout? Why the change?

File: core/thread.go [hunk 5]
```diff
@@ -184,129 +190,204 @@ func (t *Thread) UpdateSchema(hash string) error {
...
-		}
-		ends, err := t.followParent(hash)
-		if err != nil {
-			log.Warningf("failed to follow parent %s: %s", parent, err)
-			continue
-		}
-		for _, p := range ends {
-			final[p] = struct{}{}
-		}
+		wg.Add(1)
```
Comment (carsonfarmer) [inline core/thread.go]: This is sick!

File: core/thread.go [hunk 6]
```diff
@@ -478,11 +533,29 @@ func (t *Thread) commitNode(block string, additionalParents []string, updateHead
...
 		return nil, err
 	}
-	err = ipfs.AddLinkToDirectory(t.node(), dir, parentsLinkName, pnode.Cid().Hash().B58String())
+	pnodeId := pnode.Cid().Hash().B58String()
+	err = ipfs.AddLinkToDirectory(t.node(), dir, parentsLinkName, pnodeId)
 	if err != nil {
 		return nil, err
 	}
 
+	// add target
```
Comment (carsonfarmer) [inline core/thread.go]: Nice, very straightforward code here too, I like it.

File: core/invites.go (hunk 3 above)
Comment (sanderpick) [reply core/invites.go]: should really speed up joins - we can do this now because it's no longer necessary to wait for the last discovered leaves in order to properly update the HEAD

File: core/main_test.go (hunk 4 above)
Comment (sanderpick) [reply core/main_test.go]: For some reason, the random ports were causing the api health test to fail sporadically

File: cmd/files.go [hunk 7]
```diff
@@ -536,8 +536,8 @@ func FileIgnore(blockID string) error {
 // ------------------------------------
 // > files key
 
-func FileKeys(targetID string) error {
-	res, err := executeJsonCmd(http.MethodGet, "keys/"+targetID, params{}, nil)
+func FileKeys(dataID string) error {
+	res, err := executeJsonCmd(http.MethodGet, "keys/"+dataID, params{}, nil)
```
Comment (balupton) [inline cmd/files.go]: This should be documented in the changelog if this is a value change and not just a variable name change

File: cmd/main.go [hunk 8]
```diff
@@ -387,10 +387,10 @@ Stacks may include:
 	blockFilesCommand(cmds, fileListCmd, []string{"block"})
 
 	// file keys
-	fileKeysCmd := fileCmd.Command("keys", "Shows the encryption keys for each content/meta pair for the given block DAG target").Alias("key")
-	fileKeysTargetID := fileKeysCmd.Arg("block-target", "Block Target ID").Required().String()
+	fileKeysCmd := fileCmd.Command("keys", "Shows the encryption keys for each content/meta pair for the given block DAG").Alias("key")
+	fileKeysDataID := fileKeysCmd.Arg("block-data", "Block Data ID").Required().String()
```
Comment (balupton) [inline cmd/main.go]: This should be documented in the changelog if this is a value change and not just a variable name change
